src/promise/Promise.php

```diff
@@ -52,4 +53,53 @@ public function onCompletion(\Closure $onSuccess, \Closure $onFailure) : void{
 	public function isResolved() : bool{
 		return $this->shared->resolved;
 	}
+
+	/**
+	 * Returns a promise that will resolve only once all the Promises in
+	 * `$promises` have resolved. The resolution value of the returned promise
+	 * will be an array containing the resolution values of each Promises in
+	 * `$promises` indexed by the respective Promises' array keys.
+	 *
+	 * @template TPromiseValue
+	 * @template TKey of array-key
+	 * @phpstan-param array<TKey, Promise<TPromiseValue>> $promises
+	 *
+	 * @phpstan-return Promise<array<TKey, TPromiseValue>>
+	 */
+	public static function all(array $promises) : Promise {
+		/** @phpstan-var PromiseResolver<array<TKey, TPromiseValue>> $resolver */
+		$resolver = new PromiseResolver();
+		$values = [];
+		$toResolve = count($promises);
+		$continue = true;
+
+		foreach($promises as $key => $promise){
+			$promise->onCompletion(
+				function(mixed $value) use ($resolver, $key, &$toResolve, &$continue, &$values) : void{
+					$values[$key] = $value;
+
+					if(--$toResolve === 0 && $continue){
+						$resolver->resolve($values);
+					}
+				},
+				function() use ($resolver, &$continue) : void{
+					if($continue){
+						$continue = false;
+						$resolver->reject();
+					}
+				}
+			);
+
+			if(!$continue){
```

Comment:
Does `onCompletion` have any side effects? Is there any observable difference if we short circuit here, or is it just a micro optimization?

Reply:
There's no need to continue the loop if you come across a promise that's already been rejected (the only way to access this part).

Reply:
Yes, that's exactly my question. Does `onCompletion` trigger any behavior in the other promises? Would it be inconsistent if we don't call `onCompletion` there?

(This would be an inconsistency in await-generator because generators don't start executing until they are yielded, not sure if PM has any similar issues)

Reply:
I don't think I understand what you're getting at.

Reply:
I don't think so because we reject the promise as soon as one is rejected.